sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,858 @@
+                // To terminate the operator, the downstream signaled cancellation or upstream signaled error
+                // or completion. The next drain-loop iteration as a result of that signalling terminate
+                // the operator through one of the 'terminateIf*' methods followed by placing tombstone on
+                // the drain-loop.
+
+                // Below flag indicates there is no mediator currently, and there won't be one in the future,
+                // which shows the current mediator is flushed. When 'completeAfterMediatorFlush' is set
+                // and upstream signals operator completion, the flag 'noNextMediator' tells the next drain-loop
+                // iteration that it can terminate downstream gracefully.
+                mediatorHolder.noNextMediator = true;
+                return;
+            }
+
+            final Duration delay;
+            if (error == null) {
+                // TODO (anu): decide the back-off duration when no error.
+                delay = Duration.ofSeconds(1);
+            } else {
+                // error != null
+                // TODO (anu): it seems Event Hubs Processor retries only on receiver completion, not when error-ed. If true, we need a flag.
+                final int attempt = retryAttempts.incrementAndGet();
+                delay = retryPolicy.calculateRetryDelay(error, attempt);
+                if (delay == null) {
+                    // TODO (anu): log non-retriable or retry exhausted error
+                    // Invoking 'onError' to signal the next drain-loop iteration to terminate the operator
+                    // with 'error',
+                    onError(error);
+                    // once the control returns below, the next immediate drain-loop iteration (iteration
+                    // guaranteed by the above signaling) picks the signal, terminates the operator
+                    // through 'terminateIfErrored*' method and places tombstone on the drain-loop.
+                    return;
+                }
+            }
+
+            // TODO (anu): log retry attempt
+
+            mediatorHolder.nextMediatorRequestDisposable = Schedulers.boundedElastic().schedule(() -> {
+                if (cancelled) {
+                    // The downstream signaled cancellation to terminate the operator before the retry
+                    // back-off expiration.
+                    return;
+                }
+                if (done) {
+                    // The upstream signaled operator termination before the retry back-off expiration.
+                    if (completeAfterMediatorFlush) {
+                        mediatorHolder.noNextMediator = true;
+                        drain(null);
+                    }
+                    return;
+                }
+                // request upstream for a new receiver so that the resulting receiver can back new mediator.
+                upstream.request(1);
```

Comment:
Before making request, see if logging terminal state of the upstream (e.g. cancelledSubscription) is useful.